### PR TITLE
Run on Apple Silicon: --device flag, MPS loader fix, 4x faster gated delta rule

### PR DESCRIPTION
**Pull request by shreyaskarnik**
## What

Makes `reflex-serve` run on Apple Silicon through PyTorch MPS. Nothing changes on CUDA: every change is behind `device == "mps"` except the new `--device` flag, which defaults to `cuda`.

## Why each change

| Change | Problem it fixes (M3 Pro, 18 GB, torch 2.14, transformers 5.17) |
|---|---|
| `--device {cuda,mps,cpu}` on `reflex-serve` | `Engine.load` already took `device=`, the server never passed it. |
| `HF_DEACTIVATE_ASYNC_LOAD=1` on MPS | transformers' threaded weight loader copies to the device from 4 threads at once. On MPS that segfaulted (exit 139) or hung at ~400% CPU in `_materialize_copy`. Serial load takes 1.9 s for Qwen3.5-0.8B. |
| `reflex/mps.py`: matmul-only gated delta rule | Qwen3.5 is hybrid, so without Triton it runs transformers' reference `torch_chunk_gated_delta_rule`. 81% of a request was in those 18 layers, and ~30 of each layer's 37 ms was two `torch.linalg.solve_triangular` calls, which cost ~15 ms each on MPS regardless of size (1-5 ms on CPU). The systems are 64x64 unit lower triangular, so the inverse is built by recursive block inversion: six rounds of batched matmuls. |
| `PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.7` on MPS | torch's default lets MPS allocate 1.7x the recommended working set, more than physical RAM. Loading Qwen3.5-4B fp16 on 18 GB swapped ~12 GB and ended in a watchdog reboot. With the cap it raises `MPS backend out of memory` instead. Both env vars use `setdefault`, so users can override them. |

## Numbers

Same request, same answers (identical to 6 decimals on the README example), Qwen3.5-0.8B fp16 on MPS:

| Request | Before | After |
|---|---|---|
| README support-ticket example, steady state | 260 ms | 130 ms |
| 804 state + 587 question tokens, 3 questions, state cache miss | 2420 ms | 1063 ms |
| one delta-net layer call, T=728 | 35.5 ms | 8.8 ms |

## A note on the inverse

My first attempt was the shorter Neumann product `(I - L)(I + L^2)(I + L^4)...`. It matched the reference to 1e-7 on random keys and produced NaNs on real activations: neighbouring keys are nearly parallel, `L` has entries near 1, and its powers overflow. Recursive block inversion only ever forms true inverses of sub-blocks, so it is as well conditioned as forward substitution. `tests/test_mps_delta_rule.py` has a nearly-parallel-keys case that the Neumann version fails (3e38 / NaN) and this one passes (2e-7); the tests run on CPU, no GPU or model download needed.

`patch_delta_rule` swaps the module-level `torch_chunk_gated_delta_rule` in whichever modeling module the loaded model uses, so it is a no-op for pure-attention models. Non-power-of-two chunk sizes fall back to `solve_triangular`.

## Tested

- `pytest tests/test_mps_delta_rule.py tests/test_pack.py`: 12 passed; `ruff check` and `ruff format --check` clean.
- Served Qwen3.5-0.8B and Qwen3.5-2B on MPS from this branch (rebased on f84b0e8).
- Used it to play Super Mario Bros. 1-1 to the flag with Qwen3.5-2B through `fhshaik/typesafe-mario`, a few thousand requests.
- Not tested: CUDA (no hardware here; the CUDA path is unchanged), the training commands on MPS, image state on MPS.

### CODE REVIEW

File: src/reflex/engine.py [hunk 1]
```diff
@@ -264,6 +265,15 @@ def load(
...
 
+        if str(device).startswith("mps"):
+            # transformers' threaded weight loader copies to the device from 4 threads at
+            # once, which hangs or segfaults on MPS. Load serially there.
+            os.environ.setdefault("HF_DEACTIVATE_ASYNC_LOAD", "1")
+            # By default torch lets MPS allocate 1.7x the recommended working set, which is more
+            # than physical RAM on most Macs: an oversized model then swaps the machine into a
+            # watchdog reboot. Cap it so loading fails with an out-of-memory error instead.
+            os.environ.setdefault("PYTORCH_MPS_HIGH_WATERMARK_RATIO", "0.7")
+            os.environ.setdefault("PYTORCH_MPS_LOW_WATERMARK_RATIO", "0.6")
```
Comment (kshetrajna12) [inline src/reflex/engine.py]: **P2 — Keep the memory defaults compatible with existing overrides**

Could we choose these defaults together, taking any existing override into account? I reproduced a startup failure with just `PYTORCH_MPS_LOW_WATERMARK_RATIO=1.0` set: that works with PyTorch's original defaults, but the new high watermark of `0.7` makes `Engine.load(..., device="mps")` fail with `RuntimeError: invalid low watermark ratio 1`.

Respecting an existing value when choosing the other default, and validating the resulting pair, would preserve the ability to override these settings.

File: src/reflex/server.py [hunk 2]
```diff
@@ -123,6 +123,7 @@ def main(argv=None):
     ap.add_argument("--port", type=int, default=8008)
     ap.add_argument("--max-pack-tokens", type=int, default=8192)
     ap.add_argument("--dtype", default="bfloat16", choices=["bfloat16", "float16", "float32"])
+    ap.add_argument("--device", default="cuda", choices=["cuda", "mps", "cpu"])
```
Comment (kshetrajna12) [inline src/reflex/server.py:165]: **P2 — Handle MPS out-of-memory errors in the server**

Now that the server exposes MPS, could we extend the overload handler in `systemone()` to cover its allocation errors too? MPS raises a `RuntimeError`, so the existing `except torch.cuda.OutOfMemoryError` doesn't catch it. I reproduced an actual MPS allocation failure and confirmed that the endpoint returns HTTP 500 instead of the existing 529 overload response.

Recognizing MPS OOM errors specifically, while letting unrelated runtime errors propagate, would keep the overload behavior consistent across devices.

**Review by kshetrajna12 — COMMENTED** [review]
Thanks for adding Apple Silicon support! The recursive block inverse looks solid: the new kernel passed 24 numerical comparisons on MPS, and a tiny Qwen3.5 model matched before and after patching for both prefill and cache continuation.

I found two P2 edge cases around memory handling that would be good to address; I've left details and possible fixes inline.

Validation: 15 tests passed, 9 CUDA tests skipped, and lint/format checks passed. The model smoke test used a tiny randomly initialized model; I didn't run a full pretrained-model serving test.

GitHub also currently reports merge conflicts with main, so this will need an update before merging.

File: src/reflex/engine.py (hunk 1 above)
Comment (shreyaskarnik) [reply src/reflex/engine.py]: Reproduced: `PYTORCH_MPS_LOW_WATERMARK_RATIO=1.0` alone fails against the 0.7 high with `invalid low watermark ratio 1`, and works against torch's own default.

Fixed in 766b58c. `reflex.mps.memory_watermarks(env)` now chooses the pair together: a user's low raises the default high to match (low=1.0 gives high=1), a user's high lowers the default low if needed (high=0.5 gives low=0.5), `high=0` (no limit) is respected, and a valid user pair is left untouched. An invalid pair raises a `ValueError` naming both variables before torch sees it. I checked torch's actual rule first rather than guessing: `0 <= high <= 2`, and `low <= high` unless `high == 0`.

It is a pure function over a mapping, so `tests/test_mps_memory.py` covers it on CPU (10 cases, including your repro). Also ran your repro end to end: `Engine.load(..., device="mps")` with only `LOW=1.0` set now loads with high=1, low=1.0; with nothing set it is still 0.7 / 0.6.

File: src/reflex/server.py (hunk 2 above)
Comment (shreyaskarnik) [reply src/reflex/server.py:165]: Reproduced with a real allocation failure: MPS raises a plain `RuntimeError("MPS backend out of memory ...")`. It is neither `torch.cuda.OutOfMemoryError` nor `torch.OutOfMemoryError`, so the endpoint returned 500.

Fixed in e963605. `reflex.mps.is_out_of_memory()` matches that message on a `RuntimeError`; the handler turns it into the same 529, and any other `RuntimeError` still propagates (tests cover 529 for the OOM, and 500 / re-raise for an unrelated error, using a stub engine like `test_server_auth.py`).

Testing this against a real OOM turned up a second problem that also applies to the CUDA path: after the 529, every later request failed with 529 too. `empty_cache()` was being called inside the `except` block, where the traceback still holds the frames that own the tensors, so nothing was released. The handler now records the OOM, leaves the `except` block, then runs `gc.collect()` and empties the CUDA/MPS cache before raising the 529. A/B on Qwen3.5-0.8B under a tight watermark (same cap both times): cleanup inside `except`, small requests after the OOM returned 529, 529, 529; cleanup after it, 200, 200, 200. I have no CUDA hardware, so the CUDA side of that change is untested; it is the same ordering fix, but worth a look from you.

**Comment by shreyaskarnik** [comment]
Merged `main` into the branch (0797e5b) rather than rebasing, so the reviewed commit stays as it was. Two small conflicts: `engine.py` imports (`main` dropped `random`; `os` is still needed for the MPS setup) and the `Engine.load(...)` call in `server.py` (`main` moved adapter and calibration into kwargs; I kept `device=args.device`). GitHub now reports the branch as mergeable.

Both P2 items are addressed in 766b58c and e963605, details in the threads. On the merged tree: 46 passed, 9 skipped (CUDA), `ruff check` and `ruff format --check` clean. Also re-served Qwen3.5-0.8B on MPS from the branch, including the real out-of-memory case described in the second thread.

File: src/reflex/mps.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    if not 0 <= high <= 2 or low < 0 or (high != 0 and low > high):
+        raise ValueError(
+            f"invalid MPS memory watermarks: {HIGH_WATERMARK}={high:g}, {LOW_WATERMARK}={low:g}. "
+            "torch needs 0 <= high <= 2 (0 disables the limit) and low <= high."
+        )
+    chosen = {}
+    if user_high is None:
+        chosen[HIGH_WATERMARK] = f"{high:g}"
+    if user_low is None:
+        chosen[LOW_WATERMARK] = f"{low:g}"
```
Comment (kshetrajna12) [inline src/reflex/mps.py]: **P3 — Non-blocking: preserve precision when writing the watermarks**

Could we use `str(high)` / `str(low)` (or `repr`) here instead of `:g`? The default `:g` formatting rounds to six significant digits, so a valid pair can become invalid after the validation above.

For example, setting only `PYTORCH_MPS_LOW_WATERMARK_RATIO=0.7000001` computes a matching high watermark, but writes it as `0.7`. I reproduced the resulting startup failure on MPS: `RuntimeError: invalid low watermark ratio 0.7`.

Preserving the float's precision and adding a regression case for that value should cover it. This is a narrow edge case and doesn't block merging; a follow-up is fine.

**Review by kshetrajna12 — APPROVED** [review]
Thanks for the fixes and the detailed repros! Both original P2 findings are resolved, and the cleanup-ordering change checks out. This looks ready to merge to me.

I've left one small, non-blocking precision fix inline; I'm happy for it to be handled as a follow-up.

Validation on this revision: 46 tests passed, 9 CUDA tests skipped, and lint/format checks passed. Additional checks passed for all six device/--stable combinations, loading a tiny saved model through Engine.load on CPU and MPS, HTTP requests and cache reuse, and CPU/MPS probability agreement. A real MPS OOM returned 529 and subsequent requests succeeded. CUDA cleanup ordering passed a simulated-error check, though actual CUDA hardware remains untested.

File: src/reflex/mps.py (hunk 3 above)
Comment (shreyaskarnik) [reply src/reflex/mps.py]: Reproduced: with only `LOW=0.7000001` the high is computed as 0.7000001 and then written as `0.7`, so torch rejects the pair.

Fixed in b717207 rather than leaving it for a follow-up, since it is two lines. The derived values are written with `repr()`, which round trips the float exactly. Added a regression test first and watched it fail on the old code: it checks `0.7000001` and a longer value, that the chosen pair validates again when fed back in, and the mirrored case (a user high of `0.5999999` lowering the default low). Your repro now loads on MPS with high=0.7000001, low=0.7000001. One visible side effect: `LOW=1.0` now yields a high of `1.0` instead of `1`.

Full suite on this revision: 47 passed, 9 skipped; ruff check and format clean.

Comment (kshetrajna12) [reply src/reflex/mps.py]: Merge it in whenever ready

**Comment by shreyaskarnik** [comment]
do not have merge access @kshetrajna12 this is good to go.
